File: frontend/src/services/albumService.js
const API_BASE_URL = "http://localhost:8080/albums";

export async function getRandomAlbums() {
  const response = await fetch(`${API_BASE_URL}/random`);
  if (!response.ok) {
    throw new Error("Failed to fetch albums");
  }
  return response.json();
}

export async function searchAlbums(query) {
  const response = await fetch(
    `${API_BASE_URL}/search?q=${encodeURIComponent(query)}`
  );
  if (!response.ok) {
    throw new Error("Failed to search albums");
  }
  return response.json();
}

export async function getAlbumById(albumId) {
  const response = await fetch(`${API_BASE_URL}/${encodeURIComponent(albumId)}`);
  if (!response.ok) {
    throw new Error("Failed to fetch album");
  }
  return response.json();
}

export async function getSongsForAlbum(albumId) {
  const response = await fetch(
    `${API_BASE_URL}/${encodeURIComponent(albumId)}/songs`
  );
  if (!response.ok) {
    throw new Error("Failed to fetch album songs");
  }
  return response.json();
}

File: backend/src/main/java/com/example/soundwave/service/AlbumService.java
package com.example.soundwave.service;

import com.example.soundwave.dto.musicbrainz.AlbumDto;
import java.util.List;

/**
 * Service interface for album-related business logic
 *
 * This layer abstracts where artist data comes from
 * (MusicBrainz, database, cache, etc.)
 */

public interface AlbumService {

    /**
     * Returns a list of random albums
     *
     * @return list of AlbumDto
     */

    List<AlbumDto> getRandomAlbums();


    /**
     * Search for albums by name
     *
     * @param query search keyword
     * @return list of matching albums
     */

    List<AlbumDto> searchAlbums(String query);


    /**
     * Fetch a single album by ID
     *
     * @param albumId MusicBrainz album ID
     * @return AlbumDto
     */

    AlbumDto getAlbumById(String albumId);
}